lib/redlock/client.rb

```diff
@@ -83,6 +84,17 @@ class RedisInstance
           return 0
         end
       eos
+      # thanks to https://github.com/sbertrang/redis-distlock/blob/master/lib/Redis/DistLock.pm
+      # also https://github.com/sbertrang/redis-distlock/issues/2 which proposes the value-checking
+      EXTEND_SCRIPT = <<-eos
+        if redis.call( "get", KEYS[1] ) == ARGV[1] then
+          if redis.call( "set", KEYS[1], ARGV[1], "XX", "PX", ARGV[2] ) then
```

Comment:
The `"XX"` is not needed here as the line before made sure the key exists.

Instead of carrying around your `extend` variable through the various `lock` methods, I'd suggest combining the `SETNX` command for lock and the extend script into one:

```rb
def lock(resource, ttl, extend: {}, &block)
  value = extend[:value] || SecureRandom.uuid
  lock_info = try_lock_instances(resource, ttl, value)
  [...]
end

LOCK_SCRIPT = <<-eos
if redis.call("exists",KEYS[1]) == 0 or redis.call("get",KEYS[1]) == ARGV[1] then
  return redis.call("set",KEYS[1],ARGV[1],"PX",ARGV[2])
end
eos
[...]
```

Executing a script for `lock()` each time is presumable worse in terms of performance than a `SETNX` command, but I'd still prefer this for readability.